### PR TITLE
Fix #91: added OpenTSDBBucket class/datasource; some tests added as well

### LINKED ISSUE #91: OpenTSDB data source
Your help is wanted to release the OpenTSDB data source plugin in the next Loud ML release. Please check the contributing guide, and refer to datasource.py and influxdb.py for a reference data source implementation.

### PR DESCRIPTION
**Pull request by vsergeyev**
OpenTSDB module for Loud ML

 * OpenTSDBClient class handles connection to OpenTSDB database
 * OpenTSDBBucket class is capable to read and write timed data

### CODE REVIEW

**Comment by regel** [comment]
Awesome! Thanks, Volodymyr.

Some feedback inline.

File: loudml/opentsdb.py [hunk 1]
```diff
@@ -0,0 +1,364 @@
...
+
+
+class OpenTSDBBucket(Bucket):
+    """
+    OpenTSDB bucket
+    """
+
+    SCHEMA = Bucket.SCHEMA.extend({
+        Required('addr'): str,
+        Optional('use_ssl', default=False): Boolean(),
```
Comment (regel) [inline loudml/opentsdb.py]: Useful options to add:

- Username and password for basic auth if supported by OpenTSDB
- A `verify_ssl` flag, to verify the server's SSL certificate. Useful if the server is using self-signed certificates.
- Optional for this first pull-request: client certificates parameters

File: loudml/opentsdb.py [hunk 2]
```diff
@@ -0,0 +1,364 @@
...
+        **kwargs
+    ):
+        """
+        Insert data
+
+        'sync' - Whether or not to wait for the data to be flushed to storage
+        before returning the results. Used in tests
+        """
+        ts = int(make_ts(ts))
+        for k, v in data.items():
```
Comment (regel) [inline loudml/opentsdb.py]: slightly more Pythonic: 

filtered = filter(lambda item: item[1] is not None, data.items())
map
for k, v in filtered:

File: loudml/opentsdb.py [hunk 3]
```diff
@@ -0,0 +1,364 @@
...
+                self.url,
+                time_pred,
+                AGGREGATORS.get(q["aggregator"]),
+                q["down_sampler"],
+                q["metric_name"],
+                str(q["tags"])
+            )
+            # TODO: OpenTSDB is capable of running multiple subquries in
+            # one shot. Refactor it to use one request to server
+            resp = requests.get(query_url)
```
Comment (regel) [inline loudml/opentsdb.py]: See https://requests.readthedocs.io/en/latest/user/advanced/#ssl-cert-verification

**Comment by vsergeyev** [comment]
Hello @regel 

Agree with comments, thank you for reviewing it.
Applying.

Regards,
V.

File: loudml/opentsdb.py [hunk 4]
```diff
@@ -0,0 +1,364 @@
...
+    def put(self, entry):
+        """
+        Store a point in database
+        """
+        url = "%s/api/put" % self.url
+
+        if entry["sync"]:
+            url = "{}?sync".format(url)
+
+        del entry["sync"]
```
[the remark this answers is not in the document]
Comment (vsergeyev) [reply loudml/opentsdb.py]: Tryied not to delete it. OpenTSDB returns error - Unable to parse the given JSON","details":"com.fasterxml.jackson.databind.exc.UnrecognizedPropertyException: Unrecognized field "sync"

So let it be deleted

File: loudml/opentsdb.py (hunk 1 above)
Comment (vsergeyev) [reply loudml/opentsdb.py]: Agree

**Comment by vsergeyev** [comment]
It looks all comments are covered with recent commit. Sebastien, please take a look.

Regarding .circleci/config.yml - I'm not familiar with this CI system.
Will take a look, cann't promise how i will goes.
For OpenTSDB installation I used 2.3.1 from homebrew https://formulae.brew.sh/formula/opentsdb
Setup: http://localhost:4242 with default config,
and before running OpenTSDB hbase was started as well.

Regards,
V.

**Comment by regel** [comment]
> It looks all comments are covered with recent commit. Sebastien, please take a look.
> 
> Regarding .circleci/config.yml - I'm not familiar with this CI system.
> Will take a look, cann't promise how it will goes.
> For OpenTSDB installation I used 2.3.1 from homebrew https://formulae.brew.sh/formula/opentsdb
> Setup: http://localhost:4242 with default config,
> and before running OpenTSDB hbase was started as well.
> 
> Regards,
> V.

Awesome! Thank you Volodymyr

One more line is needed in `setup.py`, you need to declare the OpenTSDB class entry point.

Regarding CI, CircleCI spins up Docker containers for each commit (Or pull-request) that is pushed to the repository. Useful to run tasks like builds and tests. You can see the output of the tests directly inside the pull-request, where it says "All checks have passed".

The Docker image below is based on TSDB_VERSION 2.3.1 apparently. I'm not sure if there are other official images alternatives. One option is to use this Docker image in CI in order to run tests.

https://hub.docker.com/r/petergrace/opentsdb-docker/

Editing .circleci config should look similar to other data sources; so something that looks like this:


```
  test-opentsdb:
    docker:
      - image: circleci/python:3.6.1
      - image: petergrace/opentsdb-docker
    <<: *opentsdb_unit_test

opentsdb_unit_test: &opentsdb_unit_test
    environment:
      UNITTESTS: tests/test_opentsdb.py
      OPENTSDB_ADDR: localhost:4242
      DATABASE_ADDR: localhost:4242
    working_directory: ~/repo
    <<: *db_integration_test
```

OPENTSDB_ADDR is the environment variable read in the test file.

DATABASE_ADDR is the environment variable used to wait until the database is ready to accept connections. You can see it's configured to wait one minute max.

```
      - run:
          name: Wait for db
          command: |
            if [ ! -z "$DATABASE_ADDR" ]; then
              dockerize -wait tcp://$DATABASE_ADDR -timeout 1m
            fi
```

**Comment by vsergeyev** [comment]
Thank you Sebastien,

It looks not so bad with CircleCI :-)

Please review commits.

Regads,

V.

**Comment by regel** [comment]
> Thank you Sebastien,
> 
> It looks not so bad with CircleCI :-)
> 
> Please review commits.
> 
> Regads,
> 
> V.

Hi Volodymyr, thanks so much!

You are a true Loud ML hero 💯 I'll send some swag and will also add your name to the list of open source contributors. Please send me a private message: sebastien -at- loudml.io

Pull request approved. Have a good day!

**Pull request merged**